packages/css-parser-algorithms/src/consume/consume-component-block-function.ts

```diff
@@ -0,0 +1,621 @@
+import { CSSToken, mirrorVariantType, stringify, TokenType, isToken, TokenFunction } from '@csstools/css-tokenizer';
+import { Context } from '../interfaces/context';
+import { ComponentValueType } from '../util/component-value-type';
+
+export type ContainerNode = FunctionNode | SimpleBlockNode;
+
+export type ComponentValue = FunctionNode | SimpleBlockNode | WhitespaceNode | CommentNode | TokenNode | UnclosedSimpleBlockNode | UnclosedFunctionNode;
+
+// https://www.w3.org/TR/css-syntax-3/#consume-a-component-value
+export function consumeComponentValue(ctx: Context, tokens: Array<CSSToken>): { advance: number, node: ComponentValue } {
+	const token = tokens[0];
+	if (
+		token[0] === TokenType.OpenParen ||
+		token[0] === TokenType.OpenCurly ||
+		token[0] === TokenType.OpenSquare
+	) {
+		const r = consumeSimpleBlock(ctx, tokens);
+		return {
+			advance: r.advance,
+			node: r.node,
+		};
+	}
+
+	if (token[0] === TokenType.Function) {
+		const r = consumeFunction(ctx, tokens);
+		return {
+			advance: r.advance,
+			node: r.node,
+		};
+	}
+
+	if (token[0] === TokenType.Whitespace) {
+		const r = consumeWhitespace(ctx, tokens);
+		return {
+			advance: r.advance,
+			node: r.node,
+		};
+	}
+
+	if (token[0] === TokenType.Comment) {
+		const r = consumeComment(ctx, tokens);
+		return {
+			advance: r.advance,
+			node: r.node,
+		};
+	}
+
+	return {
+		advance: 1,
+		node: new TokenNode(token),
+	};
+}
+
+export class FunctionNode {
+	type: ComponentValueType = ComponentValueType.Function;
+
+	name: TokenFunction;
+	endToken: CSSToken;
+	value: Array<ComponentValue>;
+
+	constructor(name: TokenFunction, endToken: CSSToken, value: Array<ComponentValue>) {
+		this.name = name;
+		this.endToken = endToken;
+		this.value = value;
+	}
+
+	get nameTokenValue(): string {
+		return this.name[4].value;
```

Comment:
`4` here seems like a magic number, should it be abstracted into its own constant?

Reply:
Is indeed a bit of a magic number, but best to keep it as is for now.
I've tried to keep the tokenizer API really small so we have a bit more freedom later.
I also hope that using plain numbers is a bit faster, but this might be wishful thinking :)

All tokens are arrays, but are also fully typed. So by narrowing the type of each token to it's specific token type you can see which index holds which value/type.

0. token type
1. representation (what was found in the source)
2. start index
3. end index
4. value (what was parsed)

- representation can contain escape sequences, value will not
- value is structured and has things like `unit`

<img width="563" alt="Screenshot 2022-11-06 at 16 50 09" src="https://user-images.githubusercontent.com/11521496/200180660-e731de99-d7d8-4418-9b38-05ebe51058a3.png">

I do a agree that this is a bad API for plugin creators unless they also have a lot of knowledge about CSS tokenizing.

Maybe something like this is easier for plugin creators to use :

```ts
const myToken = [TokenType.Ident, 'foo', 0, 2, { value: 'foo' }];
const tokenType = typeFromToken(myToken); // TokenType.Ident
const tokenValue = valueFromToken(myToken); // { value: 'foo' }
```

```ts
const myToken = [TokenType.Dimension, '10px', 0, 3, { value: 10, unit: 'px' }];
const tokenType = typeFromToken(myToken); // TokenType.Dimension
const tokenValue = valueFromToken(myToken); // { value: 10, unit: 'px' }
```

Reply:
Thanks for the explanation!